File: examples/typescript/basic.ts
import { Reader } from "@mcapable/wasm";

import { sampleBytes } from "./common";

function main(): void {
  const reader = Reader.fromBytes(sampleBytes());
  const header = reader.header() as { profile: string; library: string; metadata: [string, string][] };
  console.log(`MCAP Profile: ${header.profile}`);
  if (header.metadata.length > 0) {
    console.log("Header metadata:");
    for (const [key, value] of header.metadata) {
      console.log(`  ${key}: ${value}`);
    }
  }

  console.log("\nMessages:");
  const stream = reader.messages();
  let count = 0;
  for (;;) {
    const item = stream.next();
    if (!item) break;
    const message = item as {
      channel_id: number;
      sequence: number;
      log_time: number;
      data: Uint8Array;
    };
    console.log(
      `  [${message.log_time}] channel=${message.channel_id} seq=${message.sequence} size=${message.data.length}`
    );
    if (++count >= 10) {
      console.log("  ... (showing first 10)");
      break;
    }
  }
}

main();

File: examples/cpp/basic.cc
#include "common.h"
#include "cxx.h"
#include "mcapable-cpp/src/lib.rs.h"
#include <iostream>

int main() {
  auto data = sample_bytes();
  auto reader = mcapable::reader_from_bytes(data);
  auto header = mcapable::reader_header(*reader);
  std::cout << "MCAP Profile: " << header.profile << "\n";
  if (!header.metadata.empty()) {
    std::cout << "Header metadata:\n";
    for (const auto& kv : header.metadata) {
      std::cout << "  " << kv.key << ": " << kv.value << "\n";
    }
  }

  auto stream = mcapable::reader_messages(*reader);

  mcapable::Message message{};
  size_t count = 0;
  while (mcapable::message_stream_next(*stream, message)) {
    std::cout << "  [" << message.log_time << "] channel=" << message.channel_id
              << " seq=" << message.sequence << " size=" << message.data.size() << "\n";
    if (++count >= 10) {
      std::cout << "  ... (showing first 10)\n";
      break;
    }
  }
  return 0;
}
